File: webapp/go/src/app/game.go
package main

import (
	"context"
	"fmt"
	"log"
	"math/big"
	"strconv"
	"time"

	_ "github.com/go-sql-driver/mysql"
	"github.com/gorilla/websocket"
	"github.com/jmoiron/sqlx"
)

type GameRequest struct {
	RequestID int    `json:"request_id"`
	Action    string `json:"action"`
	Time      int64  `json:"time"`

	// for addIsu
	Isu string `json:"isu"`

	// for buyItem
	ItemID      int `json:"item_id"`
	CountBought int `json:"count_bought"`
}

type GameResponse struct {
	RequestID int  `json:"request_id"`
	IsSuccess bool `json:"is_success"`
}

// 10進数の指数表記に使うデータ。JSONでは [仮数部, 指数部] という2要素配列になる。
type Exponential struct {
	// Mantissa * 10 ^ Exponent
	Mantissa int64
	Exponent int64
}

func (n Exponential) MarshalJSON() ([]byte, error) {
	return []byte(fmt.Sprintf("[%d,%d]", n.Mantissa, n.Exponent)), nil
}

type Adding struct {
	RoomName string `json:"-" db:"room_name"`
	Time     int64  `json:"time" db:"time"`
	Isu      string `json:"isu" db:"isu"`
}

type Buying struct {
	RoomName string `db:"room_name"`
	ItemID   int    `db:"item_id"`
	Ordinal  int    `db:"ordinal"`
	Time     int64  `db:"time"`
}

type Schedule struct {
	Time       int64       `json:"time"`
	MilliIsu   Exponential `json:"milli_isu"`
	TotalPower Exponential `json:"total_power"`
}

type Item struct {
	ItemID      int         `json:"item_id"`
	CountBought int         `json:"count_bought"`
	CountBuilt  int         `json:"count_built"`
	NextPrice   Exponential `json:"next_price"`
	Power       Exponential `json:"power"`
	Building    []Building  `json:"building"`
}

type OnSale struct {
	ItemID int   `json:"item_id"`
	Time   int64 `json:"time"`
}

type Building struct {
	Time       int64       `json:"time"`
	CountBuilt int         `json:"count_built"`
	Power      Exponential `json:"power"`
}

type GameStatus struct {
	Time     int64      `json:"time"`
	Adding   []Adding   `json:"adding"`
	Schedule []Schedule `json:"schedule"`
	Items    []Item     `json:"items"`
	OnSale   []OnSale   `json:"on_sale"`
}

type mItem struct {
	ItemID int   `db:"item_id"`
	Power1 int64 `db:"power1"`
	Power2 int64 `db:"power2"`
	Power3 int64 `db:"power3"`
	Power4 int64 `db:"power4"`
	Price1 int64 `db:"price1"`
	Price2 int64 `db:"price2"`
	Price3 int64 `db:"price3"`
	Price4 int64 `db:"price4"`
}

func (item *mItem) GetPower(count int) *big.Int {
	// power(x):=(cx+1)*d^(ax+b)
	a := item.Power1
	b := item.Power2
	c := item.Power3
	d := item.Power4
	x := int64(count)

	s := big.NewInt(c*x + 1)
	t := new(big.Int).Exp(big.NewInt(d), big.NewInt(a*x+b), nil)
	return new(big.Int).Mul(s, t)
}

func (item *mItem) GetPrice(count int) *big.Int {
	// price(x):=(cx+1)*d^(ax+b)
	a := item.Price1
	b := item.Price2
	c := item.Price3
	d := item.Price4
	x := int64(count)

	s := big.NewInt(c*x + 1)
	t := new(big.Int).Exp(big.NewInt(d), big.NewInt(a*x+b), nil)
	return new(big.Int).Mul(s, t)
}

func str2big(s string) *big.Int {
	x := new(big.Int)
	x.SetString(s, 10)
	return x
}

func big2exp(n *big.Int) Exponential {
	s := n.String()

	if len(s) <= 15 {
		return Exponential{n.Int64(), 0}
	}

	t, err := strconv.ParseInt(s[:15], 10, 64)
	if err != nil {
		log.Panic(err)
	}
	return Exponential{t, int64(len(s) - 15)}
}

func getCurrentTime() (int64, error) {
	var currentTime int64
	err := db.Get(&currentTime, "SELECT floor(unix_timestamp(current_timestamp(3))*1000)")
	if err != nil {
		return 0, err
	}
	return currentTime, nil
}

// 部屋のロックを取りタイムスタンプを更新する
//
// トランザクション開始後この関数を呼ぶ前にクエリを投げると、
// そのトランザクション中の通常のSELECTクエリが返す結果がロック取得前の
// 状態になることに注意 (keyword: MVCC, repeatable read).
func updateRoomTime(tx *sqlx.Tx, roomName string, reqTime int64) (int64, bool) {
	// See page 13 and 17 in https://www.slideshare.net/ichirin2501/insert-51938787
	_, err := tx.Exec("INSERT INTO room_time(room_name, time) VALUES (?, 0) ON DUPLICATE KEY UPDATE time = time", roomName)
	if err != nil {
		log.Println(err)
		return 0, false
	}

	var roomTime int64
	err = tx.Get(&roomTime, "SELECT time FROM room_time WHERE room_name = ? FOR UPDATE", roomName)
	if err != nil {
		log.Println(err)
		return 0, false
	}

	var currentTime int64
	err = tx.Get(&currentTime, "SELECT floor(unix_timestamp(current_timestamp(3))*1000)")
	if err != nil {
		log.Println(err)
		return 0, false
	}
	if roomTime > currentTime {
		log.Println("room time is future")
		return 0, false
	}
	if reqTime != 0 {
		if reqTime < currentTime {
			log.Println("reqTime is past")
			return 0, false
		}
	}

	_, err = tx.Exec("UPDATE room_time SET time = ? WHERE room_name = ?", currentTime, roomName)
	if err != nil {
		log.Println(err)
		return 0, false
	}

	return currentTime, true
}

func addIsu(roomName string, reqIsu *big.Int, reqTime int64) bool {
	tx, err := db.Beginx()
	if err != nil {
		log.Println(err)
		return false
	}

	_, ok := updateRoomTime(tx, roomName, reqTime)
	if !ok {
		tx.Rollback()
		return false
	}

	_, err = tx.Exec("INSERT INTO adding(room_name, time, isu) VALUES (?, ?, '0') ON DUPLICATE KEY UPDATE isu=isu", roomName, reqTime)
	if err != nil {
		log.Println(err)
		tx.Rollback()
		return false
	}

	var isuStr string
	err = tx.QueryRow("SELECT isu FROM adding WHERE room_name = ? AND time = ? FOR UPDATE", roomName, reqTime).Scan(&isuStr)
	if err != nil {
		log.Println(err)
		tx.Rollback()
		return false
	}
	isu := str2big(isuStr)

	isu.Add(isu, reqIsu)
	_, err = tx.Exec("UPDATE adding SET isu = ? WHERE room_name = ? AND time = ?", isu.String(), roomName, reqTime)
	if err != nil {
		log.Println(err)
		tx.Rollback()
		return false
	}

	if err := tx.Commit(); err != nil {
		log.Println(err)
		return false
	}
	return true
}

func buyItem(roomName string, itemID int, countBought int, reqTime int64) bool {
	tx, err := db.Beginx()
	if err != nil {
		log.Println(err)
		return false
	}

	_, ok := updateRoomTime(tx, roomName, reqTime)
	if !ok {
		tx.Rollback()
		return false
	}

	var countBuying int
	err = tx.Get(&countBuying, "SELECT COUNT(*) FROM buying WHERE room_name = ? AND item_id = ?", roomName, itemID)
	if err != nil {
		log.Println(err)
		tx.Rollback()
		return false
	}
	if countBuying != countBought {
		tx.Rollback()
		log.Println(roomName, itemID, countBought+1, " is already bought")
		return false
	}

	totalMilliIsu := new(big.Int)
	var addings []Adding
	err = tx.Select(&addings, "SELECT isu FROM adding WHERE room_name = ? AND time <= ?", roomName, reqTime)
	if err != nil {
		log.Println(err)
		tx.Rollback()
		return false
	}

	for _, a := range addings {
		totalMilliIsu.Add(totalMilliIsu, new(big.Int).Mul(str2big(a.Isu), big.NewInt(1000)))
	}

	var buyings []Buying
	err = tx.Select(&buyings, "SELECT item_id, ordinal, time FROM buying WHERE room_name = ?", roomName)
	if err != nil {
		log.Println(err)
		tx.Rollback()
		return false
	}
	for _, b := range buyings {
		var item mItem
		tx.Get(&item, "SELECT * FROM m_item WHERE item_id = ?", b.ItemID)
		cost := new(big.Int).Mul(item.GetPrice(b.Ordinal), big.NewInt(1000))
		totalMilliIsu.Sub(totalMilliIsu, cost)
		if b.Time <= reqTime {
			gain := new(big.Int).Mul(item.GetPower(b.Ordinal), big.NewInt(reqTime-b.Time))
			totalMilliIsu.Add(totalMilliIsu, gain)
		}
	}

	var item mItem
	tx.Get(&item, "SELECT * FROM m_item WHERE item_id = ?", itemID)
	need := new(big.Int).Mul(item.GetPrice(countBought+1), big.NewInt(1000))
	if totalMilliIsu.Cmp(need) < 0 {
		log.Println("not enough")
		tx.Rollback()
		return false
	}

	_, err = tx.Exec("INSERT INTO buying(room_name, item_id, ordinal, time) VALUES(?, ?, ?, ?)", roomName, itemID, countBought+1, reqTime)
	if err != nil {
		log.Println(err)
		tx.Rollback()
		return false
	}

	if err := tx.Commit(); err != nil {
		log.Println(err)
		return false
	}

	return true
}

func getStatus(roomName string) (*GameStatus, error) {
	tx, err := db.Beginx()
	if err != nil {
		return nil, err
	}

	currentTime, ok := updateRoomTime(tx, roomName, 0)
	if !ok {
		tx.Rollback()
		return nil, fmt.Errorf("updateRoomTime failure")
	}

	mItems := map[int]mItem{}
	var items []mItem
	err = tx.Select(&items, "SELECT * FROM m_item")
	if err != nil {
		tx.Rollback()
		return nil, err
	}
	for _, item := range items {
		mItems[item.ItemID] = item
	}

	addings := []Adding{}
	err = tx.Select(&addings, "SELECT time, isu FROM adding WHERE room_name = ?", roomName)
	if err != nil {
		tx.Rollback()
		return nil, err
	}

	buyings := []Buying{}
	err = tx.Select(&buyings, "SELECT item_id, ordinal, time FROM buying WHERE room_name = ?", roomName)
	if err != nil {
		tx.Rollback()
		return nil, err
	}

	err = tx.Commit()
	if err != nil {
		return nil, err
	}

	status, err := calcStatus(currentTime, mItems, addings, buyings)
	if err != nil {
		return nil, err
	}

	// calcStatusに時間がかかる可能性があるので タイムスタンプを取得し直す
	latestTime, err := getCurrentTime()
	if err != nil {
		return nil, err
	}

	status.Time = latestTime
	return status, err
}

func calcStatus(currentTime int64, mItems map[int]mItem, addings []Adding, buyings []Buying) (*GameStatus, error) {
	var (
		// 1ミリ秒に生産できる椅子の単位をミリ椅子とする
		totalMilliIsu = big.NewInt(0)
		totalPower    = big.NewInt(0)

		itemPower    = map[int]*big.Int{}    // ItemID => Power
		itemPrice    = map[int]*big.Int{}    // ItemID => Price
		itemOnSale   = map[int]int64{}       // ItemID => OnSale
		itemBuilt    = map[int]int{}         // ItemID => BuiltCount
		itemBought   = map[int]int{}         // ItemID => CountBought
		itemBuilding = map[int][]Building{}  // ItemID => Buildings
		itemPower0   = map[int]Exponential{} // ItemID => currentTime における Power
		itemBuilt0   = map[int]int{}         // ItemID => currentTime における BuiltCount

		addingAt = map[int64]Adding{}   // Time => currentTime より先の Adding
		buyingAt = map[int64][]Buying{} // Time => currentTime より先の Buying
	)

	for itemID := range mItems {
		itemPower[itemID] = big.NewInt(0)
		itemBuilding[itemID] = []Building{}
	}

	for _, a := range addings {
		// adding は adding.time に isu を増加させる
		if a.Time <= currentTime {
			totalMilliIsu.Add(totalMilliIsu, new(big.Int).Mul(str2big(a.Isu), big.NewInt(1000)))
		} else {
			addingAt[a.Time] = a
		}
	}

	for _, b := range buyings {
		// buying は 即座に isu を消費し buying.time からアイテムの効果を発揮する
		itemBought[b.ItemID]++
		m := mItems[b.ItemID]
		totalMilliIsu.Sub(totalMilliIsu, new(big.Int).Mul(m.GetPrice(b.Ordinal), big.NewInt(1000)))

		if b.Time <= currentTime {
			itemBuilt[b.ItemID]++
			power := m.GetPower(itemBought[b.ItemID])
			totalMilliIsu.Add(totalMilliIsu, new(big.Int).Mul(power, big.NewInt(currentTime-b.Time)))
			totalPower.Add(totalPower, power)
			itemPower[b.ItemID].Add(itemPower[b.ItemID], power)
		} else {
			buyingAt[b.Time] = append(buyingAt[b.Time], b)
		}
	}

	for _, m := range mItems {
		itemPower0[m.ItemID] = big2exp(itemPower[m.ItemID])
		itemBuilt0[m.ItemID] = itemBuilt[m.ItemID]
		price := m.GetPrice(itemBought[m.ItemID] + 1)
		itemPrice[m.ItemID] = price
		if 0 <= totalMilliIsu.Cmp(new(big.Int).Mul(price, big.NewInt(1000))) {
			itemOnSale[m.ItemID] = 0 // 0 は 時刻 currentTime で購入可能であることを表す
		}
	}

	schedule := []Schedule{
		Schedule{
			Time:       currentTime,
			MilliIsu:   big2exp(totalMilliIsu),
			TotalPower: big2exp(totalPower),
		},
	}

	// currentTime から 1000 ミリ秒先までシミュレーションする
	for t := currentTime + 1; t <= currentTime+1000; t++ {
		totalMilliIsu.Add(totalMilliIsu, totalPower)
		updated := false

		// 時刻 t で発生する adding を計算する
		if a, ok := addingAt[t]; ok {
			updated = true
			totalMilliIsu.Add(totalMilliIsu, new(big.Int).Mul(str2big(a.Isu), big.NewInt(1000)))
		}

		// 時刻 t で発生する buying を計算する
		if _, ok := buyingAt[t]; ok {
			updated = true
			updatedID := map[int]bool{}
			for _, b := range buyingAt[t] {
				m := mItems[b.ItemID]
				updatedID[b.ItemID] = true
				itemBuilt[b.ItemID]++
				power := m.GetPower(b.Ordinal)
				itemPower[b.ItemID].Add(itemPower[b.ItemID], power)
				totalPower.Add(totalPower, power)
			}
			for id := range updatedID {
				itemBuilding[id] = append(itemBuilding[id], Building{
					Time:       t,
					CountBuilt: itemBuilt[id],
					Power:      big2exp(itemPower[id]),
				})
			}
		}

		if updated {
			schedule = append(schedule, Schedule{
				Time:       t,
				MilliIsu:   big2exp(totalMilliIsu),
				TotalPower: big2exp(totalPower),
			})
		}

		// 時刻 t で購入可能になったアイテムを記録する
		for itemID := range mItems {
			if _, ok := itemOnSale[itemID]; ok {
				continue
			}
			if 0 <= totalMilliIsu.Cmp(new(big.Int).Mul(itemPrice[itemID], big.NewInt(1000))) {
				itemOnSale[itemID] = t
			}
		}
	}

	gsAdding := []Adding{}
	for _, a := range addingAt {
		gsAdding = append(gsAdding, a)
	}

	gsItems := []Item{}
	for itemID, _ := range mItems {
		gsItems = append(gsItems, Item{
			ItemID:      itemID,
			CountBought: itemBought[itemID],
			CountBuilt:  itemBuilt0[itemID],
			NextPrice:   big2exp(itemPrice[itemID]),
			Power:       itemPower0[itemID],
			Building:    itemBuilding[itemID],
		})
	}

	gsOnSale := []OnSale{}
	for itemID, t := range itemOnSale {
		gsOnSale = append(gsOnSale, OnSale{
			ItemID: itemID,
			Time:   t,
		})
	}

	return &GameStatus{
		Adding:   gsAdding,
		Schedule: schedule,
		Items:    gsItems,
		OnSale:   gsOnSale,
	}, nil
}

func serveGameConn(ws *websocket.Conn, roomName string) {
	log.Println(ws.RemoteAddr(), "serveGameConn", roomName)
	defer ws.Close()

	status, err := getStatus(roomName)
	if err != nil {
		log.Println(err)
		return
	}

	err = ws.WriteJSON(status)
	if err != nil {
		log.Println(err)
		return
	}

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	chReq := make(chan GameRequest)

	go func() {
		defer cancel()
		for {
			req := GameRequest{}
			err := ws.ReadJSON(&req)
			if err != nil {
				log.Println(err)
				return
			}

			select {
			case chReq <- req:
			case <-ctx.Done():
				return
			}
		}
	}()

	ticker := time.NewTicker(500 * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case req := <-chReq:
			log.Println(req)

			success := false
			switch req.Action {
			case "addIsu":
				success = addIsu(roomName, str2big(req.Isu), req.Time)
			case "buyItem":
				success = buyItem(roomName, req.ItemID, req.CountBought, req.Time)
			default:
				log.Println("Invalid Action")
				return
			}

			if success {
				// GameResponse を返却する前に 反映済みの GameStatus を返す
				status, err := getStatus(roomName)
				if err != nil {
					log.Println(err)
					return
				}

				err = ws.WriteJSON(status)
				if err != nil {
					log.Println(err)
					return
				}
			}

			err := ws.WriteJSON(GameResponse{
				RequestID: req.RequestID,
				IsSuccess: success,
			})
			if err != nil {
				log.Println(err)
				return
			}
		case <-ticker.C:
			status, err := getStatus(roomName)
			if err != nil {
				log.Println(err)
				return
			}

			err = ws.WriteJSON(status)
			if err != nil {
				log.Println(err)
				return
			}
		case <-ctx.Done():
			return
		}
	}
}

File: webapp/nodejs/Game.js
const bigint = require('bigint')
const MItem = require('./MItem')
const Exponential = require('./Exponential')

class Game {
  constructor(roomName, pool) {
    this.roomName = roomName
    this.pool = pool
  }

  async getStatus () {
    const connection = await this.pool.getConnection()
    await connection.beginTransaction()

    try {
      const currentTime = await this.updateRoomTime(connection, 0)
      const mItems = {}
      const [items] = await connection.query('SELECT * FROM m_item')
      for (let item of items) {
        mItems[item.item_id] = new MItem(item)
      }
      const [addings] = await connection.query('SELECT time, isu FROM adding WHERE room_name = ?', [this.roomName])
      const [buyings] = await connection.query('SELECT item_id, ordinal, time FROM buying WHERE room_name = ?', [this.roomName])
      await connection.commit()
      connection.release()

      const status = this.calcStatus(currentTime, mItems, addings, buyings)

      // calcStatusに時間がかかる可能性があるので タイムスタンプを取得し直す
      const latestTime = await this.getCurrentTime()
      status.time = latestTime

      return status

    } catch (e) {
      await connection.rollback()
      connection.release()
      throw e
    }
  }

  async addIsu (reqIsu, reqTime) {
    try {
      const connection = await this.pool.getConnection()
      await connection.beginTransaction()

      try {
        await this.updateRoomTime(connection, reqTime)
        await connection.query('INSERT INTO adding(room_name, time, isu) VALUES (?, ?, \'0\') ON DUPLICATE KEY UPDATE isu=isu', [this.roomName, reqTime])

        const [[{ isu }]] = await connection.query('SELECT isu FROM adding WHERE room_name = ? AND time = ? FOR UPDATE', [this.roomName, reqTime])
        const newIsu = reqIsu.add(bigint(isu))

        await connection.query('UPDATE adding SET isu = ? WHERE room_name = ? AND time = ?', [newIsu.toString(), this.roomName, reqTime])
        await connection.commit()
        connection.release()
        return true

      } catch (e) {
        await connection.rollback()
        connection.release()
        throw e
      }
    } catch (e) {
      console.error(e)
      return false
    }
  }

  async buyItem (itemId, countBought, reqTime) {
    try {
      const connection = await this.pool.getConnection()
      await connection.beginTransaction()

      try {
        await this.updateRoomTime(connection, reqTime)
        const [[{ countBuying }]] = await connection.query('SELECT COUNT(*) as countBuying FROM buying WHERE room_name = ? AND item_id = ?', [this.roomName, itemId])
        if (parseInt(countBuying, 10) != countBought) {
          throw new Error(`roomName=${this.roomName}, itemId=${itemId} countBought+1=${countBought+1} is already bought`)
        }

        let totalMilliIsu = bigint('0')
        const [addings] = await connection.query('SELECT isu FROM adding WHERE room_name = ? AND time <= ?', [this.roomName, reqTime])
        for (let { isu } of addings) {
          totalMilliIsu = totalMilliIsu.add(bigint(isu).mul(bigint('1000')))
        }

        const [buyings] = await connection.query('SELECT item_id, ordinal, time FROM buying WHERE room_name = ?', [this.roomName])
        for (let b of buyings) {
          let [[mItem]] = await connection.query('SELECT * FROM m_item WHERE item_id = ?', [b.item_id])
          let item = new MItem(mItem)
          let cost = item.getPrice(parseInt(b.ordinal, 10)).mul(bigint('1000'))
          totalMilliIsu = totalMilliIsu.sub(cost)
          if (parseInt(b.time, 10) <= reqTime) {
            let gain = item.getPower(parseInt(b.ordinal, 10)).mul(bigint('' + (reqTime - parseInt(b.time, 10))))
            totalMilliIsu = totalMilliIsu.add(gain)
          }
        }

        const [[mItem]] = await connection.query('SELECT * FROM m_item WHERE item_id = ?', [itemId])
        const item = new MItem(mItem)
        const need = item.getPrice(countBought + 1).mul(bigint('1000'))
        if (totalMilliIsu.cmp(need) < 0) {
          throw new Error('not enough')
        }

        await connection.query('INSERT INTO buying(room_name, item_id, ordinal, time) VALUES(?, ?, ?, ?)', [this.roomName, itemId, countBought + 1, reqTime])
        await connection.commit()
        connection.release()
        return true

      } catch (e) {
        await connection.rollback()
        connection.release()
        throw e
      }
    } catch (e) {
      console.error(e)
      return false
    }
  }

  // 部屋のロックを取りタイムスタンプを更新する
  //
  // トランザクション開始後この関数を呼ぶ前にクエリを投げると、
  // そのトランザクション中の通常のSELECTクエリが返す結果がロック取得前の
  // 状態になることに注意 (keyword: MVCC, repeatable read).
  async updateRoomTime (connection, reqTime) {
    // See page 13 and 17 in https://www.slideshare.net/ichirin2501/insert-51938787
    await connection.query('INSERT INTO room_time(room_name, time) VALUES (?, 0) ON DUPLICATE KEY UPDATE time = time', [this.roomName])
    const [[{ time }]] = await connection.query('SELECT time FROM room_time WHERE room_name = ? FOR UPDATE', [this.roomName])
    const [[{ currentTime }]] = await connection.query('SELECT floor(unix_timestamp(current_timestamp(3))*1000) AS currentTime')
    if (parseInt(time, 10) > parseInt(currentTime, 10)) {
      throw new Error('room time is future')
    }
    if (reqTime !== 0) {
      if (reqTime < parseInt(currentTime, 10)) {
        throw new Error('reqTime is past')
      }
    }

    await connection.query('UPDATE room_time SET time = ? WHERE room_name = ?', [currentTime, this.roomName])
    return currentTime
  }

  calcStatus (currentTime, mItems, addings, buyings) {
    // 1ミリ秒に生産できる椅子の単位をミリ椅子とする
    let totalMilliIsu = bigint('0')
    let totalPower    = bigint('0')

    const itemPower    = {} // ItemID => Power
    const itemPrice    = {} // ItemID => Price
    const itemOnSale   = {} // ItemID => OnSale
    const itemBuilt    = {} // ItemID => BuiltCount
    const itemBought   = {} // ItemID => CountBought
    const itemBuilding = {} // ItemID => Buildings
    const itemPower0   = {} // ItemID => currentTime における Power
    const itemBuilt0   = {} // ItemID => currentTime における BuiltCount

    const addingAt = {} // Time => currentTime より先の Adding
    const buyingAt = {} // Time => currentTime より先の Buying

    for (let itemId in mItems) {
      itemPower[itemId] = bigint('0')
      itemBuilding[itemId] = []
    }

    for (let a of addings) {
      // adding は adding.time に isu を増加させる
      if (a.time <= currentTime) {
        totalMilliIsu = totalMilliIsu.add(bigint(a.isu).mul(bigint('1000')))
      } else {
        addingAt[a.time] = a
      }
    }

    for (let b of buyings) {
      // buying は 即座に isu を消費し buying.time からアイテムの効果を発揮する
      itemBought[b.item_id] = itemBought[b.item_id] ? itemBought[b.item_id] + 1 : 1
      const m = mItems[b.item_id]
      totalMilliIsu = totalMilliIsu.sub(m.getPrice(b.ordinal).mul(bigint('1000')))

      if (b.time <= currentTime) {
        itemBuilt[b.item_id] = itemBuilt[b.item_id] ? itemBuilt[b.item_id] + 1 : 1
        const power = m.getPower(itemBought[b.item_id])
        totalMilliIsu = totalMilliIsu.add(power.mul(bigint(currentTime - b.time)))
        totalPower = totalPower.add(power)
        itemPower[b.item_id] = itemPower[b.item_id].add(power)
      } else {
        buyingAt[b.time] = buyingAt[b.time] || []
        buyingAt[b.time].push(b)
      }
    }

    for (let itemId in mItems) {
      const m = mItems[itemId]
      itemPower0[m.itemId] = this.big2exp(itemPower[m.itemId])
      itemBuilt0[m.itemId] = itemBuilt[m.itemId]
      const price = m.getPrice((itemBought[m.itemId] || 0) + 1)
      itemPrice[m.itemId] = price
      if (0 <= totalMilliIsu.cmp(price.mul(bigint('1000')))) {
        itemOnSale[m.itemId] = 0 // 0 は 時刻 currentTime で購入可能であることを表す
      }
    }

    const schedule = [
      {
        time:        currentTime,
        milli_isu:   this.big2exp(totalMilliIsu),
        total_power: this.big2exp(totalPower),
      }
    ]

    // currentTime から 1000 ミリ秒先までシミュレーションする
    for (let t = currentTime + 1; t <= currentTime + 1000; t++) {
      totalMilliIsu = totalMilliIsu.add(totalPower)
      let updated = false

      // 時刻 t で発生する adding を計算する
      if (addingAt[t]) {
        let a = addingAt[t]
        updated = true
        totalMilliIsu = totalMilliIsu.add(bigint(a.isu).mul(bigint('1000')))
      }

      // 時刻 t で発生する buying を計算する
      if (buyingAt[t]) {
        updated = true
        const updatedID = {}
        for (let b of buyingAt[t]) {
          const m = mItems[b.item_id]
          updatedID[b.item_id] = true
          itemBuilt[b.item_id] = itemBuilt[b.item_id] ? itemBuilt[b.item_id] + 1 : 1
          const power = m.getPower(b.ordinal)
          itemPower[b.item_id] = itemPower[b.item_id].add(power)
          totalPower = totalPower.add(power)
        }
        for (let id in updatedID) {
          itemBuilding[id].push({
            time:        t,
            count_built: itemBuilt[id],
            power:       this.big2exp(itemPower[id]),
          })
        }
      }

      if (updated) {
        schedule.push({
          time:        t,
          milli_isu:   this.big2exp(totalMilliIsu),
          total_power: this.big2exp(totalPower),
        })
      }

      // 時刻 t で購入可能になったアイテムを記録する
      for (let itemId in mItems) {
        if (typeof itemOnSale[itemId] !== 'undefined') {
          continue;
        }
        if (0 <= totalMilliIsu.cmp(itemPrice[itemId].mul(bigint('1000')))) {
          itemOnSale[itemId] = t
        }
      }
    }

    const gsAdding = []
    for (let a of Object.values(addingAt)) {
      gsAdding.push(a)
    }

    const gsItems = []
    for (let itemId in mItems) {
      gsItems.push({
        item_id:      parseInt(itemId, 10),
        count_bought: itemBought[itemId] || 0,
        count_built:  itemBuilt0[itemId] || 0,
        next_price:   this.big2exp(itemPrice[itemId]),
        power:        itemPower0[itemId],
        building:     itemBuilding[itemId],
      })
    }

    const gsOnSale = []
    for (let itemId in itemOnSale) {
      let t = itemOnSale[itemId]
      gsOnSale.push({
        item_id: parseInt(itemId, 10),
        time:    t,
      })
    }

    return {
      time:      0,
      adding:    gsAdding,
      schedule:  schedule,
      items:     gsItems,
      on_sale:   gsOnSale,
    }
  }

  async getCurrentTime () {
    try {
      const [[{currentTime}]] = await this.pool.query('SELECT floor(unix_timestamp(current_timestamp(3))*1000) AS currentTime')
      return parseInt(currentTime, 10)
    } catch (e) {
      console.error(e)
      return 0
    }
  }

  big2exp (n) {
    const s = n.toString()
    if (s.length <= 15) {
      return new Exponential({
        mantissa: n.toNumber(),
        exponent: 0,
      })
    }

    const t = parseInt(s.slice(0, 15), 10)
    return new Exponential({
      mantissa: t,
      exponent: s.length - 15,
    })
  }
}

module.exports = Game
